### DISCUSSION TITLE
Minimal changes needed for new non-static DP and AppContext for CSLA 6

### DISCUSSION THREAD

**Discussion by swegele** [MarimerLLC/csla#2670] [Questions] [answered]
Let me see if I understand the implications of the answer for #2632 ... and to ask about what I _think might be the most minimal changes_ to my business objects (and others like me) to get up to v6.

In that ticket, @rockfordlhotka said:
> Static factory methods are no longer an option.
> Your choices are to inject (or create) a DataPortal<T> instance and use it, or put your static factory methods into a factory class (so they become instance methods of that class) and inject the DataPortal<T> into that class.

He further explained that the whole thing stems from no longer being able to get a static reference to ApplicationContext which is required to get a DataPortal.

**Question One** - So is it correct that if I 
1.  get (via DI) a reference to ApplicationContext (within my UI I guess)
2.  pass that into my root object's static Create/Fetch/Delete etc. methods  
3.  then "inside" my business object, I can create an instance of the DataPortal<T> and proceed to call the DP like before 
(this would make it so that no one outside the business objects has to know anything about the DataPortal)

So this is what I had previously:
```
public static AgencyEdit NewAgency()
{
     var newObj = DataPortal.Create();
     newObj.BusinessRules.CheckRules();
     return newObj;
}
```

that could now become this:

```
public static AgencyEdit NewAgency(ApplicationContext appContext)
{
     var dp = new DataPortal<AgencyEdit>(appContext, ??);
     var newObj = dp.Create();
     newObj.BusinessRules.CheckRules();
     return newObj;
}
```

Notice the "??".  I'm not sure what is that 2nd parameter on the DataPortal constructor:  Csla.DataPortalClient.IDataPortalProxy
What is the IDataPortalProxy and how would I properly create/use it to get a DataPortal instance? 

Finally, I do realize that if I can DI up an instance of ApplicationContext, then I could simply also get a proper DataPortal object...but for some reason that seems to make less sense to me.  Maybe just my old twisted brain that will finally click eventually.

Thank you 1000 times for all your hard work on v6 Rocky.  I just gotta get my mind wrapped around this big change.

**Comment by rockfordlhotka** [comment, marked as answer]
I have some example code in my blog post about the [CSLA 6 prerelease](https://blog.lhotka.net/2021/11/29/CSLA-6-First-Prerelease).

Also, there is some discussion in that post's discussion area that includes (for WinForms/WPF apps) a way to expose a `GetDataPortal` method from the `App` object.

**Reply by rockfordlhotka** [reply]
The question really is whether you want to "modernize" your UI code to use and rely on DI like in aspnetcore and Blazor. Or you can minimize changes to your existing code and largely ignore DI by exposing some global methods that provide access to key services (like creating a data portal).

You can't use that global method concept in aspnetcore or Blazor, because those UI frameworks already rely on DI to function, and such global methods wouldn't work. But they are a decent hack to allow limited use of DI in older UI frameworks, hopefully minimizing changes to your existing codebase.

**Reply by swegele** [reply]
When upgrading to CSLA 6, I actually have the received the rare luxury (permission from management) to re-skin the UI layer from ASP.Net WebForms to Server Side Blazor.  YAY!!  I will be able to modernize and use DI as you suggest.

But, regardless I will still have static factory methods on many of my root business objects because many of those require various other objects/parameters at their creation time.  Here is a simple example:
```
[Serializable()]
public class CaseEdit : RootBusinessBase<CaseEdit>
{
        public static CaseEdit NewCase(Agency.AgencyInfo agency, string agencyCaseNumber)
        {
            //Call DataPortal sending a criteria
            ...

        }

        [Create]
        private void DP_Create(NewCaseCriteria criteria)
        {
            //Do DAL stuff
            ...
        }

```

As you well know, there has never been an intelli-sense (design-time) way for consumers of our business objects (other developers and testers) to see what are the matching parameters required on the server side DataPortal methods ([Create], [Fetch], [Delete]).  I don't want those consumers to have to go digging in code to see what parameters to send from the UI or tests.  It really helps them, knowing they can use intelli-sense to find and call the static method that is the one they are interested in, and pass in what is asked for...they know it will work.  
So, based on DI requirements, now the trick is for me to ask for what is now (and perhaps will someday be) needed.  Some variation of one of the below solutions should work?

**Option 1 - At the most basic and most flexible level, I could require IServiceProvider as a parameter **
1. Add IServiceProvider into the static NewCase method  
2. Use that, inside the NewCase method to get instances of ApplicationContext and DataPortal<T>

This solves the issue you brought up which is that to get a DataPortal<T> you need ApplicationContext...and for that you need IServiceProvider.  But it absolutely assumes that DI is in use.

```
        public static CaseEdit NewCase(IServiceProvider serviceProvider, Agency.AgencyInfo agency, string agencyCaseNumber)
        {
            //Use sp here to get required services (DataPortal/ApplicationContext/IDataPortalProxy/whatever else)
            ...

        }

```

**Option 2 - At a little higher level I could require parameters of ApplicationContext and DataPortal<CaseEdit> **
1. Add these two required parameters to the static NewCase method  
2. Use that, inside the NewCase 

```
public static CaseEdit NewCase(ApplicationContext appContext, DataPortal<CaseEdit> dp, Agency.AgencyInfo agency, string agencyCaseNumber)
{
     ...
}
```

**Option 3 - Happy medium??  Simply require parameter of ApplicationContext and IDataPortalProxy **
Use these to create the DataPortal<T>

Thoughts?  God I hope I'm not out in left field here.  I actually feel like I'm starting to get it so that would be depressing :-)

**Reply by swegele** [reply]
I can't help notice that you have this on ApplicationContext.

```
internal IServiceProvider CurrentServiceProvider
    {
      get => _serviceProvider;
      set => _serviceProvider = value;
    }
```

Maybe this would be related to #2658 as a helpful thing to expose as public.  So if I were inside an actual instance of one of my business objects I could get a hold of:
1. ApplicationContext
2. GetDataPortal method
3. ApplicationContext.CurrentServiceProvider

```
public IServiceProvider CurrentServiceProvider
    {
      get => _serviceProvider;
      internal set => _serviceProvider = value;
    }
```

??

**Reply by rockfordlhotka** [reply]
Interesting. The ultimate decision-maker on why we implemented this massive change to adopt DI is because of the unique way server-side Blazor uses DI. It is the first UI framework that works the way it does, and so _everything_ depends on correct use of DI in SSB. You can't really cheat like in WinForms or WPF.

SSB uses a DI scope (scoped provider) for isolation on a per-session basis. Because there's no way to get the current scoped provider from a static method (or any other code that is running outside of DI), you can't get access to any DI-based services. Interestingly enough, the current user is a DI-based service, and so you can't even find out who's logged into your app in a static method.

This is why ApplicationContext switched from a set of static methods to an instance: everything in ApplicationContext that is per-user or unique to a user or session _used to be_ accessible via static methods, and now can _only be accessed via DI_.

Your best bet, imo, is to take your static factory methods and move them to a factory class, so they are instance methods of the factory class. Then you can inject the factory into your Blazor pages/components.

```c#
  public class CustomerEditFactory
  {
    public CustomerEditFactory(IDataPortal<CustomerEdit> portal)
    {
      Portal = portal;
    }

    private IDataPortal<CustomerEdit> Portal { get; set; }

    public async Task<CustomerEdit> GetCustomerEditAsync(int id)
    {
      return await Portal.FetchAsync(id);
    }
  }
```

**Reply by swegele** [reply]
I've been heavily researching SSB and you are right...it's unique and tough to get one's mind wrapped around (especially when trying to remember that your javascript is actually running on the server!) ... but it's seriously cool and I'm really looking forward to seeing how this cross over from web forms goes.  

I like your CustomerEditFactory idea.  It's definitely a more "native DI-ish" way to get what is needed.  But I've always liked "less indirection/abstraction" so I'd go for less classes but that is just a preference.  For example, thats why I've kept my DAL implementation inside my server side dataportal methods.  One stop simple shop...but I know it's not the most testable and breaks single responsibility bla bla.  I know.

Anyway, let's assume I use DI in the SSB component/page to get a DataPortal<T>

Can you see how it would be helpful for me to be able to get from that one thing to the other things...if you exposed the property getters as public?
Example:
make the DataPortal<T>.ApplicationContext getter public

and

make the ApplicationContext.ServiceProvider getter public

I know I can already use DI to get those various things individually/separately...but then I'd have to pass three things around instead of just one.  Bottom line, it is still DI that is providing what is needed.  Just kind of like using DI to get the big russian doll which already has all the littler dolls inside of it.

**Reply by rockfordlhotka** [reply]
Those properties aren't public because I don't want to be responsible for opening a path for people to abuse and misuse DI. That's just not a reputation that would be helpful to CSLA. Instead, I'm trying to retain as much backward compatibility I can, while still supporting established patterns for building apps in environments like SSB.

**Reply by swegele** [reply]
OK I'll assume I'm veering off the proper breadcrumb path and keep coming back to directly using DI to get the things I need one way or another.  Appreciate the dialog and ideas.  Will keep learning and playing.
